fix: @W-19837693: Handling multiple versions migration and assessment for Omniscript/Flexcards

sf-kishore-kurri opened the pull request:
Changes:
- Handle of multiple versions having same name pre/post clean up with allVersions
- Added logic to check for unique and duplicate based on earlier versions
- Added test cases

### What does this PR do?

### What issues does this PR fix or reference?

sf-kishore-kurri committed:
fix: @W-19837693: Handling multiple versions migration and assessment for Omniscript and Flexcards

Changes:
- Handle of multiple versions having same name pre/post clean up with allVersions
- Added logic to check for unique and duplicate based on earlier versions
- Added test cases

sf-aastha-paruthi commented on src/migration/flexcard.ts:
```diff
@@ -783,8 +802,16 @@ export class CardMigrationTool extends BaseMigrationTool implements MigrationToo
 
-      if (uniqueNames.has(uniqueCheckName)) {
-        this.setRecordErrors(card, this.messages.getMessage('duplicatedCardName', [uniqueCheckName]));
+      if (uniqueNames.has(uniqueCheckName) || dupFlexCardNames.has(transformedCard['Name'])) {
```
The changes in this file appear similar. Please move to common function.

sf-aastha-paruthi commented on src/migration/flexcard.ts:
```diff
@@ -230,11 +235,23 @@ export class CardMigrationTool extends BaseMigrationTool implements MigrationToo
+      }
+      if (uniqueNames.has(uniqueCleanedName)) {
+        flexCardAssessmentInfo.warnings.push(this.messages.getMessage('duplicateCardNameMessage', [uniqueCleanedName]));
+      } else {
```
The if or else will be set to true when multiple versions of flexcards are there.
A - v1
A - v2
So warning will get populated which is not intended.

sf-kishore-kurri replied:
Discussed with @sf-aastha-paruthi and resolved the scenario, but FC delete case is missing

sf-kishore-kurri committed:
fix: @W-19837693: Address the code review for delete case of OS and FC

Changes:
- Handle the delete use case with multiple versions of OS and FC
- Added test cases
- Updated the messages

TODO:
- Update the labels for CX review

sf-aastha-paruthi commented on src/migration/flexcard.ts:
```diff
@@ -231,11 +236,29 @@ export class CardMigrationTool extends BaseMigrationTool implements MigrationToo
       assessmentStatus = getUpdatedAssessmentStatus(assessmentStatus, 'Needs manual intervention');
     }
+    // Check for naming conflict: different original names cleaning to same name
+    else if (this.allVersions && dupFlexCardNames.has(cleanedName)) {
```
The code is duplicate. Can be separated to a separate function.

sf-kishore-kurri replied:
with migration, we set the status on record whereas in assessment its only message with status

sf-aastha-paruthi commented on test/migration/omniscript-allversions-duplicate.test.ts:
```diff
@@ -314,12 +314,230 @@ describe('OmniScript Multiple Versions Duplicate Check', () => {
+      expect(resultAS3.migrationStatus).to.equal('Needs manual intervention');
+      expect(resultAS3.name).to.equal('A_Type_English_3');
+      expect(resultAS3.warnings.length).to.be.greaterThan(0);
+      expect(resultAS3.warnings.some((w) => w.includes('Lower version duplicate'))).to.be.true;
```
Why is it being populated with lower version duplicate issue?

sf-kishore-kurri replied:
We don't have exact duplicate exists so bucketed under lower version duplicate case, if we had version1/2 which had marked for duplicate

sf-aastha-paruthi replied:
A %s with name \"%s\" will not be migrated because lower version of same %s will be marked as duplicate, which could lead to conflicts during migration.

But the message gives the impression that, the omniscript has some lower version, which has been marked as duplicate.

Can you please elaborate what the message trying to convey?
